lib/src/multi_sig/multi_sig_builder.ts

```
@@ -0,0 +1,331 @@
+
+    // 32 bytes message for session creation. E.g. fund tx hash
+    public message?: string;
+
+    // current session, could be complete or incomplete
+    private session?: string;
+
+    // self nonce commitment
+    public nonceCommitment?: string;
+
+    // local cache for the already added co-signer's commitments
+    private coSignerNonceCommitments?: Array<string> = [];
+
+    // current nonce
+    public nonce?: Buffer;
+
+    // local cache for the already added co-signer's nonces
+    private coSignerNonces?: Array<Buffer> = [];
+
+    // session's own partial signature, will be passed to co-signers
+    public partialSignature?: Buffer;
+
+    // local cache for the already added co-signer's signatures
+    private coSignerSignatures?: Array<Buffer> = [];
+
+    // session's final signature
+    public signature?: string;
+
+    constructor(
+        selfKeyPair: { privateKey?: Buffer; publicKey?: Buffer },
+        otherSignerPublicKeys: Array<PublicKey>,
+        network: NetworkEnum,
+        requiredSignerNumber: number,
+    ) {
+        ow(
+            selfKeyPair,
+            ow.object.exactShape({
+                privateKey: ow.buffer,
+                publicKey: ow.buffer,
+            }),
+        );
+        ow(otherSignerPublicKeys, ow.array.minLength(1));
+        ow(otherSignerPublicKeys, ow.array.ofType(ow.buffer));
+        ow(network, ow.string);
+        ow(requiredSignerNumber, ow.number.uint32);
+        ow(requiredSignerNumber, ow.number.greaterThanOrEqual(1));
+
+        this.keyPair = selfKeyPair;
+        this.network = network;
+        this.signerPublicKeys = otherSignerPublicKeys
+            .slice()
+            .concat([selfKeyPair.publicKey!]);
+        this.requiredSignerNumber = requiredSignerNumber;
+    }
+
+    /**
+     * create multiSig address for receiving fund
+     * @returns {multiSigAddress}
+     * @memberof MultiSigBuilder
+     */
+    public createMultiSigAddress(): string | undefined {
```

In what case will it return undefined?

---

if `native.multiSig.createAddress` throw an error? And the catch flow for every native call could be tedious I think, what do you think

---

When it throws an Error, the control flow "jumps away" instead of returning a value I think? 🤔

But yes I think the current native call error handling is not very well implemented. Perhaps we should have an extra layer between the JS code and native code.